rpc/src/paging.rs

```diff
@@ -0,0 +1,82 @@
+//! Pagination-related data structures for the Tendermint RPC.
+
+use crate::Error;
+use serde::{Deserialize, Serialize};
+use std::convert::{TryFrom, TryInto};
+use std::str::FromStr;
+
+/// Pagination control for those RPC client methods supporting pagination.
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub enum Paging {
+    /// No explicit options set - use whatever the endpoint's defaults are.
+    Default,
+    /// Try to automatically fetch all pages' data.
+    All,
+    /// Fetch a specific page's data.
+    Specific {
+        /// The number of the page to fetch.
+        page_number: PageNumber,
+        /// The number of items to fetch per page.
+        per_page: PerPage,
+    },
+}
+
+/// A page number in paginated RPC responses.
+#[derive(Debug, Clone, Copy, Serialize, Deserialize, Eq, PartialEq, PartialOrd, Ord)]
+pub struct PageNumber(usize);
+
+impl FromStr for PageNumber {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let raw = i64::from_str(s).map_err(|e| Error::client_internal_error(e.to_string()))?;
+        let raw_usize: usize = raw.try_into().map_err(|_| {
+            Error::client_internal_error(format!("page number out of range: {}", raw))
+        })?;
+        raw_usize.try_into()
+    }
+}
+
+impl std::fmt::Display for PageNumber {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl TryFrom<usize> for PageNumber {
```

Comment:
Shouldn't this be a `From` instance?

Reply:
Yes, you're right - it's a vestige of an earlier time where I tried to enforce the 1-based numbering system. But just doing that, given that I had to write custom de/serializers, resulted in a significant amount more code, which I'm trying hard to avoid.

I'll change this to `From` :+1: